bin/pasta_prepare_evaluation.py

```diff
@@ -343,33 +353,229 @@ def _load_responses_dict(msg_id, response_list):
     log.info("Total clusters found by pasta: {}".format(len(clusters)))
 
 
+def pre_process_response_data(config):
+    # Load responses dict into dataframe, preliminary processing, indexing
+
+    with open(config.f_responses_pkl, 'rb') as handle:
+        data = pickle.load(handle)
+
+    response_df = pd.DataFrame(data)
+    log.info("Done reading input in pandas dataframe")
+
+    # Convert set to list
+    response_df['upstream'] = response_df['upstream'].map(list)
+
+    response_df.index.name = "idx"
+
+    response_df.fillna({'patch_id': '_'}, inplace=True)
+    log.info("Filled NA for patch_id")
+
+    response_df.set_index(['patch_id'], append=True, inplace=True)
+    log.info("Done setting index for response_df")
+
+    # Denormalize
+    df_melt_responses = pd.melt(response_df.responses.apply(pd.Series).reset_index(),
+                                id_vars=['idx', 'patch_id'],
+                                value_name='responses').sort_index()
+
+    df_melt_responses.drop('variable', axis=1, inplace=True)
+
+    log.info("melt_responses_shape {}".format(df_melt_responses.shape))
+
+    df_denorm_responses = flat_table.normalize(df_melt_responses, expand_dicts=True, expand_lists=True)
+    df_denorm_responses.drop('index', axis=1, inplace=True)
+    df_denorm_responses.drop_duplicates(inplace=True)
+    log.info("Computed de-normalized responses, writing to disk...")
+
+    df_denorm_responses.to_csv(config.f_denorm_responses, index=False)
+    log.info("Processed responses!")
+
+    df_melt_upstream = pd.melt(response_df.upstream.apply(pd.Series).reset_index(),
+                               id_vars=['idx', 'patch_id'],
+                               value_name='upstream').sort_index()
+
+    df_melt_upstream.drop('variable', axis=1, inplace=True)
+    df_melt_upstream.drop_duplicates(inplace=True)
+
+    df_melt_upstream.to_csv(config.f_denorm_upstream, index=False)
+    log.info("Processed upstream!")
+    log.info("Finished processing")
+
+
+def merge_pre_processed_response_dfs(config):
+    def try_literal_eval(s):
+        try:
+            return literal_eval(s)
+        except ValueError:
+            return s
+
+    def _get_message_field(msg, field):
+        if not (np.all(pd.isnull(msg))):
+            return email.message_from_bytes(msg)[field]
+        else:
+            return None
+
+    dd1 = dd.read_csv(config.f_denorm_responses, blocksize=1e9, dtype={"idx ": "int32", "patch_id ": "category",
```

Comment:
So if we introduce dask + pandas as a dependency, we should use it elsewhere as well, where we can benefit. E.g., the startup of PaStA is pretty slow when loading the indexes. I'll check if we can use pandas for reading the csvs - so if we introduce pands, it'd be great if there's a chance to boost or improve other things with it

Reply:
BTW, what influence does the blocksize have?

Reply:
> BTW, what influence does the blocksize have?

This is to specify how you chunk your csv

Reply:
> So if we introduce dask + pandas as a dependency, we should use it elsewhere as well, where we can benefit. E.g., the startup of PaStA is pretty slow when loading the indexes. I'll check if we can use pandas for reading the csvs - so if we introduce pands, it'd be great if there's a chance to boost or improve other things with it

It is a good idea to test this.

Reply:
> > BTW, what influence does the blocksize have?
> 
> This is to specify how you chunk your csv

chunk for what? As a test, I tried to, for example, parse indexes with panda (it's indeed faster), but I didn't see any influence of the blocksize, as there's no parallelisation or sth. similar (afaict).

Reply:
> > > BTW, what influence does the blocksize have?
> > 
> > 
> > This is to specify how you chunk your csv
> 
> chunk for what? As a test, I tried to, for example, parse indexes with panda (it's indeed faster), but I didn't see any influence of the blocksize, as there's no parallelisation or sth. similar (afaict).

A dask dataframe is a made up of pandas dataframes (indexed and split). Blocksize affects these partitions when csv is ingested as a dask dataframe. It has to be chosen to balance good resource usage - cpus, memory, and scheduling overhead. I played around with it when my data was gigantic. I cannot however justify this value, may be it is an artifact of some of the trials I made on some dataframe along the way. May be the defaults are good enough. We have to check this.

Reply:
Ah, ok. Thanks for the explanation! Had similar issues with the parallelisation/Pool library of python: you can specify a worksize and a chunksize. The weight of those values have massive influence on memory consumption, however, one should have a look into a magic crystal ball to find sane values :-)